File: android/src/main/kotlin/com/fluttergaplessloop/BpmDetector.kt
package com.fluttergaplessloop

import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round
import kotlin.math.sqrt

/**
 * BPM/Tempo detector using the Ellis (2007) energy onset beat tracking algorithm.
 *
 * No Android framework dependencies — pure Kotlin, fully thread-safe (no shared state).
 *
 * Algorithm:
 *  1. Mix stereo to mono; compute RMS energy per 512-sample frame (256-sample hop).
 *  2. Positive half-wave rectified energy derivative → onset strength envelope.
 *  3. Weighted autocorrelation (Gaussian prior at 120 BPM) → BPM estimate.
 *  4. Dynamic programming forward pass → globally consistent beat sequence.
 */
internal object BpmDetector {

    private const val FRAME_SIZE          = 512
    private const val HOP_SIZE            = 256
    private const val MIN_BPM             = 60.0
    private const val MAX_BPM             = 180.0
    private const val TEMPO_PRIOR_MEAN    = 120.0
    private const val TEMPO_PRIOR_SIGMA   = 30.0
    private const val DP_LAMBDA           = 100.0
    private const val MIN_DURATION_SECS   = 2.0

    /** Output of BPM detection. */
    data class BpmResult(
        val bpm: Double,          // 0.0 if detection failed/skipped
        val confidence: Double,   // [0.0, 1.0]
        val beats: List<Double>   // timestamps in seconds
    )

    /**
     * Detects BPM and beat timestamps from decoded PCM audio.
     *
     * @param pcm         Interleaved float samples, range [-1.0, 1.0].
     * @param sampleRate  Sample rate in Hz (e.g. 44100, 48000).
     * @param channelCount 1 (mono) or 2 (stereo).
     * @return [BpmResult] with bpm=0.0 and empty beats if audio is too short or silent.
     */
    fun detect(pcm: FloatArray, sampleRate: Int, channelCount: Int): BpmResult {
        val totalFrames     = pcm.size / channelCount
        val durationSeconds = totalFrames.toDouble() / sampleRate

        if (durationSeconds < MIN_DURATION_SECS) {
            return BpmResult(0.0, 0.0, emptyList())
        }

        // Stage 1: Onset strength envelope
        val mono  = mixToMono(pcm, channelCount)
        val onset = computeOnsetStrength(mono)

        val maxOnset = onset.maxOrNull() ?: 0f
        if (maxOnset == 0f) return BpmResult(0.0, 0.0, emptyList())
        for (i in onset.indices) onset[i] /= maxOnset   // normalize to [0, 1]

        // Stage 2: BPM estimation via weighted autocorrelation
        val lagMin = (sampleRate.toDouble() / (MAX_BPM / 60.0) / HOP_SIZE).toInt().coerceAtLeast(1)
        val lagMax = (sampleRate.toDouble() / (MIN_BPM / 60.0) / HOP_SIZE).toInt()
            .coerceAtMost(onset.size / 2)

        if (lagMin >= lagMax) return BpmResult(0.0, 0.0, emptyList())

        val ac      = autocorrelate(onset, lagMin, lagMax, sampleRate)
        val bestIdx = ac.indices.maxByOrNull { ac[it] }
            ?: return BpmResult(0.0, 0.0, emptyList())
        val actualLag    = bestIdx + lagMin
        val estimatedBpm = 60.0 * sampleRate / (actualLag.toDouble() * HOP_SIZE)
        val confidence   = ac[bestIdx].toDouble().coerceIn(0.0, 1.0)

        // Stage 3: DP beat sequence
        val period = round(sampleRate.toDouble() / (estimatedBpm / 60.0) / HOP_SIZE).toInt()
        val beats  = trackBeats(onset, period, sampleRate)

        return BpmResult(estimatedBpm, confidence, beats)
    }

    // ─── Private Helpers ──────────────────────────────────────────────────────

    private fun mixToMono(pcm: FloatArray, channelCount: Int): FloatArray {
        if (channelCount == 1) return pcm.copyOf()
        val frames = pcm.size / channelCount
        return FloatArray(frames) { f ->
            var sum = 0f
            for (ch in 0 until channelCount) sum += pcm[f * channelCount + ch]
            sum / channelCount
        }
    }

    private fun computeOnsetStrength(mono: FloatArray): FloatArray {
        val nFrames = (mono.size - FRAME_SIZE) / HOP_SIZE + 1
        val rms = FloatArray(nFrames) { f ->
            val start = f * HOP_SIZE
            var sumSq = 0.0
            for (i in start until start + FRAME_SIZE) {
                val s = mono[i].toDouble()
                sumSq += s * s
            }
            sqrt(sumSq / FRAME_SIZE).toFloat()
        }
        // Positive half-wave rectified derivative
        val onset = FloatArray(nFrames)
        for (f in 1 until nFrames) onset[f] = max(0f, rms[f] - rms[f - 1])
        return onset
    }

    private fun autocorrelate(
        onset: FloatArray, lagMin: Int, lagMax: Int, sampleRate: Int
    ): FloatArray {
        val n    = onset.size
        val nLags = lagMax - lagMin + 1
        val ac   = FloatArray(nLags)

        for (i in 0 until nLags) {
            val lag   = lagMin + i
            val count = n - lag
            if (count <= 0) continue

            var sum = 0.0
            for (t in 0 until count) sum += onset[t].toDouble() * onset[t + lag].toDouble()
            val normalized = sum / count

            // Gaussian tempo prior: centre at 120 BPM, σ = 30 BPM
            val bpm    = 60.0 * sampleRate / (lag.toDouble() * HOP_SIZE)
            val z      = (bpm - TEMPO_PRIOR_MEAN) / TEMPO_PRIOR_SIGMA
            val weight = exp(-0.5 * z * z)
            ac[i] = (normalized * weight).toFloat()
        }

        // Normalize to [0, 1] so peak = confidence
        val maxAc = ac.maxOrNull() ?: return ac
        if (maxAc > 0f) for (i in ac.indices) ac[i] /= maxAc
        return ac
    }

    private fun trackBeats(onset: FloatArray, period: Int, sampleRate: Int): List<Double> {
        val n = onset.size
        if (period <= 0 || n < period) return emptyList()

        val score = FloatArray(n) { -Float.MAX_VALUE }
        val prev  = IntArray(n) { -1 }

        // Seed: first two periods may contain the first beat
        for (t in 0 until min(period * 2, n)) score[t] = onset[t]

        val halfPeriod = period / 2
        val twoPeriod  = period * 2
        for (t in period until n) {
            for (d in halfPeriod..twoPeriod) {
                val b = t - d
                if (b < 0 || score[b] == -Float.MAX_VALUE) continue
                val logRatio = ln(d.toDouble() / period)
                val penalty  = (DP_LAMBDA * logRatio * logRatio).toFloat()
                val candidate = score[b] + onset[t] - penalty
                if (candidate > score[t]) { score[t] = candidate; prev[t] = b }
            }
        }

        // Backtrack from best frame in last period
        var best = n - 1
        for (t in max(0, n - period) until n) if (score[t] > score[best]) best = t

        val beatFrames = mutableListOf<Int>()
        var t = best
        while (t >= 0 && score[t] > -Float.MAX_VALUE) {
            beatFrames.add(0, t)
            t = prev[t]
        }

        // Convert to seconds, drop micro-fade region (first 5ms)
        val microFadeSeconds = 0.005
        return beatFrames
            .map { f -> f.toDouble() * HOP_SIZE / sampleRate }
            .filter { ts -> ts >= microFadeSeconds }
    }
}

File: ios/Classes/BpmDetector.swift
#if os(iOS)
import AVFoundation
import Foundation

// MARK: - BpmResult

/// The result of BPM/tempo detection.
public struct BpmResult {
    /// Estimated tempo in beats per minute. `0.0` if detection failed or was skipped.
    public let bpm: Double
    /// Confidence in [0.0, 1.0]. Values above 0.5 indicate reliable detection.
    public let confidence: Double
    /// Beat timestamps in seconds from the start of the file.
    public let beats: [Double]
}

// MARK: - BpmDetector

/// BPM/Tempo detector using the Ellis (2007) energy onset beat tracking algorithm.
///
/// All methods are pure functions — no shared mutable state, fully thread-safe.
///
/// Algorithm:
///  1. Mix stereo to mono; compute RMS energy per 512-sample frame (256-sample hop).
///  2. Positive half-wave rectified energy derivative → onset strength envelope.
///  3. Weighted autocorrelation (Gaussian prior at 120 BPM) → BPM estimate.
///  4. Dynamic programming forward pass → globally consistent beat sequence.
public enum BpmDetector {

    private static let frameSize: Int       = 512
    private static let hopSize: Int         = 256
    private static let minBpm               = 60.0
    private static let maxBpm               = 180.0
    private static let tempoPriorMean       = 120.0
    private static let tempoPriorSigma      = 30.0
    private static let dpLambda             = 100.0
    private static let minDurationSeconds   = 2.0

    /// Detects BPM and beat timestamps from a decoded `AVAudioPCMBuffer`.
    ///
    /// - Parameter buffer: Float PCM buffer with `floatChannelData`.
    /// - Returns: `BpmResult`. bpm=0 if audio is too short or silent.
    public static func detect(buffer: AVAudioPCMBuffer) -> BpmResult {
        let sampleRate   = buffer.format.sampleRate
        let channelCount = Int(buffer.format.channelCount)
        let totalFrames  = Int(buffer.frameLength)
        let duration     = Double(totalFrames) / sampleRate

        guard duration >= minDurationSeconds,
              let channelData = buffer.floatChannelData else {
            return BpmResult(bpm: 0, confidence: 0, beats: [])
        }

        // Stage 1: Onset strength envelope
        let mono  = mixToMono(channelData: channelData, channelCount: channelCount,
                              frameCount: totalFrames)
        var onset = computeOnsetStrength(mono: mono, frameCount: totalFrames)

        guard let maxOnset = onset.max(), maxOnset > 0 else {
            return BpmResult(bpm: 0, confidence: 0, beats: [])
        }
        for i in onset.indices { onset[i] /= maxOnset }

        // Stage 2: BPM estimation via weighted autocorrelation
        let lagMin = max(1, Int((sampleRate / (maxBpm / 60.0)) / Double(hopSize)))
        let lagMax = min(onset.count / 2, Int((sampleRate / (minBpm / 60.0)) / Double(hopSize)))

        guard lagMin < lagMax else {
            return BpmResult(bpm: 0, confidence: 0, beats: [])
        }

        let (ac, bestLagIdx) = autocorrelate(onset: onset, lagMin: lagMin,
                                             lagMax: lagMax, sampleRate: sampleRate)
        let actualLag    = bestLagIdx + lagMin
        let estimatedBpm = 60.0 * sampleRate / (Double(actualLag) * Double(hopSize))
        let confidence   = min(1.0, Double(ac[bestLagIdx]))

        // Stage 3: DP beat sequence
        let period = Int((sampleRate / (estimatedBpm / 60.0) / Double(hopSize)).rounded())
        let beats  = trackBeats(onset: onset, period: period, sampleRate: sampleRate)

        return BpmResult(bpm: estimatedBpm, confidence: confidence, beats: beats)
    }

    // MARK: - Private Helpers

    private static func mixToMono(
        channelData: UnsafePointer<UnsafeMutablePointer<Float>>,
        channelCount: Int,
        frameCount: Int
    ) -> [Float] {
        if channelCount == 1 {
            return Array(UnsafeBufferPointer(start: channelData[0], count: frameCount))
        }
        return (0 ..< frameCount).map { f in
            var sum: Float = 0
            for ch in 0 ..< channelCount { sum += channelData[ch][f] }
            return sum / Float(channelCount)
        }
    }

    private static func computeOnsetStrength(mono: [Float], frameCount: Int) -> [Float] {
        let nFrames = (frameCount - frameSize) / hopSize + 1
        var rms = [Float](repeating: 0, count: nFrames)
        for f in 0 ..< nFrames {
            let start = f * hopSize
            var sumSq: Float = 0
            for i in start ..< start + frameSize { let s = mono[i]; sumSq += s * s }
            rms[f] = sqrtf(sumSq / Float(frameSize))
        }
        var onset = [Float](repeating: 0, count: nFrames)
        for f in 1 ..< nFrames { onset[f] = max(0, rms[f] - rms[f - 1]) }
        return onset
    }

    private static func autocorrelate(
        onset: [Float], lagMin: Int, lagMax: Int, sampleRate: Double
    ) -> ([Float], Int) {
        let n     = onset.count
        let nLags = lagMax - lagMin + 1
        var ac    = [Float](repeating: 0, count: nLags)

        for i in 0 ..< nLags {
            let lag   = lagMin + i
            let count = n - lag
            guard count > 0 else { continue }
            var sum: Double = 0
            for t in 0 ..< count { sum += Double(onset[t]) * Double(onset[t + lag]) }
            let normalized = sum / Double(count)
            let bpm = 60.0 * sampleRate / (Double(lag) * Double(hopSize))
            let z   = (bpm - tempoPriorMean) / tempoPriorSigma
            ac[i]   = Float(normalized * exp(-0.5 * z * z))
        }

        let maxAc = ac.max() ?? 0
        if maxAc > 0 { for i in ac.indices { ac[i] /= maxAc } }
        let bestIdx = ac.indices.max(by: { ac[$0] < ac[$1] }) ?? 0
        return (ac, bestIdx)
    }

    private static func trackBeats(onset: [Float], period: Int, sampleRate: Double) -> [Double] {
        let n = onset.count
        guard period > 0, n >= period else { return [] }

        var score = [Float](repeating: -.greatestFiniteMagnitude, count: n)
        var prev  = [Int](repeating: -1, count: n)
        for t in 0 ..< min(period * 2, n) { score[t] = onset[t] }

        let halfPeriod = period / 2
        let twoPeriod  = period * 2
        for t in period ..< n {
            for d in halfPeriod ... twoPeriod {
                let b = t - d
                guard b >= 0, score[b] > -.greatestFiniteMagnitude else { continue }
                let logR    = log(Double(d) / Double(period))
                let penalty = Float(dpLambda * logR * logR)
                let cand    = score[b] + onset[t] - penalty
                if cand > score[t] { score[t] = cand; prev[t] = b }
            }
        }

        var best = n - 1
        for t in max(0, n - period) ..< n where score[t] > score[best] { best = t }

        var beatFrames = [Int]()
        var t = best
        while t >= 0 && score[t] > -.greatestFiniteMagnitude {
            beatFrames.insert(t, at: 0)
            t = prev[t]
        }

        let microFadeSeconds = 0.005
        return beatFrames
            .map { Double($0) * Double(hopSize) / sampleRate }
            .filter { $0 >= microFadeSeconds }
    }
}
#endif // os(iOS)
